subsys/net/lib/sockets/sockets.c

```diff
@@ -2825,3 +2950,158 @@ static bool inet_is_supported(int family, int type, int proto)
 NET_SOCKET_REGISTER(af_inet46, NET_SOCKET_DEFAULT_PRIO, AF_UNSPEC,
 		    inet_is_supported, zsock_socket_internal);
 #endif /* CONFIG_NET_NATIVE */
+
+
+#if defined(CONFIG_NET_SOCKETS_OBJ_CORE)
+struct k_obj_core_stats_desc sock_obj_type_stats_desc = {
+	.raw_size = sizeof(struct sock_obj_type_raw_stats),
+	.raw = sock_obj_stats_raw,
+	.reset = sock_obj_core_stats_reset,
+	.disable = NULL,    /* Stats gathering is always on */
+	.enable = NULL,     /* Stats gathering is always on */
+};
+
+static void set_fields(struct sock_obj *obj, int fd,
+		       struct net_socket_register *reg,
+		       int family, int type, int proto)
+{
+	obj->fd = fd;
+	obj->socket_family = family;
+	obj->socket_type = type;
+	obj->socket_proto = proto;
+	obj->reg = reg;
+	obj->creator = k_current_get();
+	obj->create_time = sys_clock_tick_get();
+}
+
+static int sock_obj_core_alloc(int fd, struct sock_obj **obj,
+			       struct net_socket_register *reg,
+			       int family, int type, int proto)
+{
+	int ret, i;
+
+	k_mutex_lock(&sock_obj_mutex, K_FOREVER);
+
+	/* Try not to allocate already closed sockets so that we
+	 * can see historical data.
+	 */
+	for (i = 0; i < ARRAY_SIZE(sock_objects); i++) {
```

Comment:
This repeated loop kind of bugs me, maybe we could handle the slot finding in a single iteration to avoid code duplication? Something like:
```c
	*obj = NULL;

	for (i = 0; i < ARRAY_SIZE(sock_objects); i++) {
		if (sock_objects[i].fd < 0) {
			if (sock_objects[i].init_done == false) {
				/* Return immediately if found unused entry */
				*obj = &sock_objects[i];
				break;
			} else if (*obj == NULL) {
				*obj = &sock_objects[i];
			}
		}
	}

	if (*obj == NULL) {
		ret = -ENOENT;
	} else {
		set_fields(*obj, fd, reg, family, type, proto);
		ret = 0;
	}
```

Reply:
Indeed, this looks much better.